### PR TITLE
Redirect `babel` parser to `babel-flow`

### PR DESCRIPTION
**Pull request by fisker**
Redirect `babel` parser to `babel-flow`, when

- starts with `// @flow`
- `*.js.flow` file

Context

- https://github.com/prettier/prettier/issues/8204#issuecomment-622578788
- https://github.com/prettier/prettier/issues/9036#issuecomment-679175487

----

- [ ] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/pr-XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: src/language-js/parser-babel.js [hunk 1]
```diff
@@ -59,12 +59,21 @@ function resolvePluginsConflict(
 }
 
 function createParse(parseMethod, ...pluginCombinations) {
-  return (text, parsers, opts) => {
+  return (text, parsers, opts = {}) => {
+    if (
+      opts.parser === "babel" &&
+      (text.startsWith("// @flow") ||
+        (opts.filepath && opts.filepath.endsWith(".js.flow")))
```
Comment (evilebottnawi) [inline src/language-js/parser-babel.js:66]: Maybe just `.flow`?

Comment (fisker) [reply src/language-js/parser-babel.js:66]: https://github.com/prettier/prettier/blob/4382203c9c34baff459ebbee6de006a9932cfb06/tests_integration/__tests__/__snapshots__/support-info.js.snap#L315-L341

Comment (fisker) [reply src/language-js/parser-babel.js:66]: Only `.js.flow` listted.

**Comment by fisker** [comment]
Changes(Not quotes around `1`)

<details>
<summary>No `// @flow`</summary>

**Prettier pr-9071**
[Playground link](https://deploy-preview-9071--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEBDABAXncAOiARnyVQF8QAaECABxgEtoBnZUVAJ3YgHcAFDhCxSoANt1QBPFlQBG7VGADWcGAGUaC+lADmyGOwCucKnAC2MuABNLVgDKodB1NrgAxCO1OoYDHcjQGMBCUIAAWMKYiAOqh9PBMGmBwqoJx9ABucRL+YEzSIFpMcOwwvPLaXsgAZqJFVABWTAAeAELySiqqqKZwtlpw1bXGII1NqlraInAAigYQ8IMidSAa7EXs-jKoFiIhNOxaMFH0ljChyAAcAAxU+xBFUfI0-vtw6+kDVACOc-BltEI0EwALRQOBWKwhdhwH70aFlZyVJA1JbDIqmeh6QxoiZTWbzAbIoZUGDbY6nc5IABMJPk9BEEwAwhBTEiQG8AKwhAxFAAq2yEKOW6SMAEkoDZYKowAc6ABBCWqGASKaLIqkUhAA)
```sh
--parser babel
```

**Input:**
```jsx
a = {"1":a}
```

**Output:**
```jsx
a = { 1: a };

```
</details>

<details>
<summary>With `// @flow` (This PR)</summary>

**Prettier pr-9071**
[Playground link](https://deploy-preview-9071--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEB6VACAAgMwDYQDuAOlAIYYC8GwxIAjHUmQL4gA0IEADjAJbQAzslBkATmKIAFcQmEoyeQmQCewzgCMxZMAGs4MAMrcdfKAHNkMMQFc4nOAFsNcACau3AGTIWbZc3AAYhBijmQw-BbIIGQ2MBAcIAAWMI54AOpJfPCCJmBwhnLZfABu2SrRYILqIGaCcGIwUtrmYcg4ivWcAFaCAB4AQtp6BoZkjnCeZnDtnfYgvX2GZuZ4cACKNhDws3hdICZi9WLRGmQueIncYmYw6XyuMEnIABwADJzXEPXp2tzR1zgxxKM04AEctvBmjx5DFBABaKBwNxuRJiOAQvjo5r+NpIDp7eb1Rx8Ky2IkrNabbYzfFzTgwc73R7PJAAJgZ2j4eBWAGEII48SAgQBWRI2eoAFXO8gJ+xKdgAklAPLBDGAbrwAIIqwwwFRrXb1FgsIA)
```sh
--parser babel
```

**Input:**
```jsx
// @flow
a = {"1":a}
```

**Output:**
```jsx
// @flow
a = { "1": a };

```
</details>


<details>
<summary>With `// @flow` (Stable)</summary>

**Prettier 2.1.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEB6VACAAgMwDYQDuAOlAIYYC8GwxIAjHUmQL4gA0IEADjAJbQAzslBkATmKIAFcQmEoyeQmQCewzgCMxZMAGs4MAMrcdfKAHNkMMQFc4nOAFsNcACau3AGTIWbZc3AAYhBijmQw-BbIIGQ2MBAcIAAWMI54AOpJfPCCJmBwhnLZfABu2SrRYILqIGaCcGIwUtrmYcg4ivWcAFaCAB4AQtp6BoZkjnCeZnDtnfYgvX2GZuZ4cACKNhDws3hdICZi9WLRGmQueIncYmYw6XyuMEnIABwADJzXEPXp2tzR1zgxxKM04AEctvBmjx5DFBABaKBwNxuRJiOAQvjo5r+NpIDp7eb1Rx8Ky2IkrNabbYzfFzTgwc73R7PJAAJgZ2j4eBWAGEII48SAgQBWRI2eoAFXO8gJ+xKdgAklAPLBDGAbrwAIIqwwwFRrXb1FgsIA)
```sh
--parser babel
```

**Input:**
```jsx
// @flow
a = {"1":a}
```

**Output:**
```jsx
// @flow
a = { 1: a };

```
</details>

File: src/language-js/parser-babel.js (hunk 1 above)
Comment (evilebottnawi) [reply src/language-js/parser-babel.js:66]: Fine :+1:

File: src/language-js/parser-babel.js [hunk 2]
```diff
@@ -59,12 +59,21 @@ function resolvePluginsConflict(
 }
 
 function createParse(parseMethod, ...pluginCombinations) {
-  return (text, parsers, opts) => {
+  return (text, parsers, opts = {}) => {
+    if (
+      opts.parser === "babel" &&
+      (text.startsWith("// @flow") ||
```
Comment (thorn0) [inline src/language-js/parser-babel.js:65]: > You can also use the form `/* @flow */` for the flag as well.

https://flow.org/en/docs/usage/#toc-prepare-your-code-for-flow

Also we should support `@noflow`: https://flow.org/en/docs/config/options/#toc-max-header-tokens-integer

Also, can't we reuse the logic (`jest-docblock`) used for `--require-pragma` here?

Comment (fisker) [reply src/language-js/parser-babel.js]: All done.

**Comment by thorn0** [comment]
Looks good enough, but an entry for changelog is needed.

**Comment by thorn0** [comment]
Turns out Babel already does that! Check this:

**Prettier 2.1.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEB6VACAAgMwDYQDuAOlDgDwAqAfABRwCUIANCBAA4wCW0AzsqACGAJ2FEACiIT8UgvIUEBPfqwBGwwWADWcGAGV2mrlADmyGMICucVnAC2quABMnzgDKDTlwSbgAxCGE7QRhuU2QQQUsYCBYQAAsYOzwAdXiueF5DMDg9aQyuADcMxQiwXhUQY144YRhxDRNg5Bw5GtYAK14ADwAhDW1dPUE7ODdjOBa2mxAu7r1jEzw4AEVLCHgpvHaQQ2Ea4QjVQUc8OPZhYxgUricYeOQADgAGVguIGpSNdgiLuAPCpNWABHdbwBocGSRXgAWigcGczjiwjgoK4KIaPmaSFa2xmNTsXHMVnxi2Waw2kxx01YMBONzuDyQACZaRouHhFgBhCB2bEgf4AVjilhqlBOMlxO0K1gAklBXLA9GBLpwAIIKvQwRTLLY1AC++qAA)
```sh
--parser babel
```

**Input:**
```jsx
// @flow
f<T>(e)
```

**Output:**
```jsx
// @flow
f<T>(e);

```

Against this:

**Prettier 2.1.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEB6VACAAgDzzgHSgDMAeAFQD4AKOAShABoQIAHGAS2gGdlQBDAE6CIAdwAKQhLxT8ANqP4BPXswBGg-mADWcGAGVWWjlADmyGIICucZnAC2auABNnLgDL8zV-qbgAxCEF7fhhOM2QQfisYCCYQAAsYezkAdQSOeG4jMDh9aUyOADdMpUiwblUQE244QRhxTVMQ5GJ5WuYAK24cACFNHT19fns4dxM4VvbbEG6cfRNTOTgARSsIeCm5DpAjQVrBSLV+Jzl41kETGFSOZxgE5AAOAAZmC4ha1M1WSIu4A6Kk2YAEd1vBGmwZFFuABaKBwFwueKCOCgjgoxq+FpINrbGa1ewcCzWfGLZZrDaTHHTZgwE43O4PJAAJlpmg4ckWAGEIPZsSB-gBWeJWWrkE4yXE7Io2ACSUDcsH0YEu7AAggr9DAlMstrUAL76oA)
```sh
--parser babel
```

**Input:**
```jsx
// @xxxx
f<T>(e)
```

**Output:**
```jsx
// @xxxx
f < T > e;

```

**Comment by fisker** [comment]
Yes , but print didn't switch to `babel-flow`

**Comment by thorn0** [comment]
Yes, we still need this change then.

I noticed Babel's pragma detection logic is different from this PR. E.g.

**Prettier pr-9071**
[Playground link](https://deploy-preview-9071--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEB6VACAJgMx7gHSnQwAEcAbCAdyJwB4AVAPgApgMCQAmLpDbhgC+AShAAaEBAAOMAJbQAzslABDAE7qaABQ0JlKVRWqqAnsskAjdarABrODADK023KgBzZDHUBXOJJwALaWcFhYYQAyqp6+qh5wAGIQ6kGqMPKeyCCqvjAQEiAAFjBBFADqRXLwiq5gcE761XIAbtWm2WCKFiDuinDqMNo2HmnIOEb9kgBWigAeAEI29o5OqkFwke5w45MBILNzTu4eFHAAir4Q8LsUUyCu6v3q2ZaqoRSF0uruMOVyWBgRWQAA4AAySb4QfrlGzSbLfODPFo7SQARyu8GGMgMOUUAFooHAwmFCuo4Bi5OThvExkgJnd9v0gnJvH4mSczpdrjt6XtJDB3v9AcCkNwBTY5BQTgBhCBBOkgJEAVkKvn6jHeBgZ9xa-gAklAIrAnGAfrIAIJGpwwUxnW79IRCIA)
```sh
--parser babel
```

**Input:**
```jsx
// dffdf
// @flow
f<T>({ "2": 2 })
```

**Output:**
```jsx
// dffdf
// @flow
f<T>({ 2: 2 });

```

**Comment by fisker** [comment]
To align with the babel logic, we need get the comments first? But it's possible we can't parse with `babel` parser, I see two options

1, try `babel-flow` parser first
2, use https://github.com/jonschlinkert/parse-comments to parse comment first

But they all seems expansive cost

**Comment by thorn0** [comment]
How about something like `/@(no)?flow)\b/.test(text.slice(0, getNextNonSpaceNonCommentCharacterIndexWithStartIndex(text, 0)))`?

**Comment by fisker** [comment]
I can try this.

**Comment by fisker** [comment]
Works good, thank you!

File: src/language-js/parser-babel.js [hunk 3]
```diff
@@ -58,13 +62,42 @@ function resolvePluginsConflict(
   return combinations;
 }
 
+// https://github.com/babel/babel/pull/7934/files#diff-a739835084910b0ee3ea649df5a4d223R67
+const FLOW_PRAGMA_REGEX = /\*?\s*@(?:no)?flow\b/;
```
Comment (thorn0) [inline src/language-js/parser-babel.js:66]:
```suggestion
const FLOW_PRAGMA_REGEX = /@(?:no)?flow\b/;
```
This part at the beginning doesn't make any difference.

Comment (fisker) [reply src/language-js/parser-babel.js:66]: Right, how  about `(?:^|[*\s])`

Comment (thorn0) [reply src/language-js/parser-babel.js:66]: We don't need anything in front of `@` to align with Babel. Babel is okay even with `// wolf@flow`.

Comment (fisker) [reply src/language-js/parser-babel.js:66]: You are right, there is many problems in this regex. `someone@flow.com` 😄

Comment (thorn0) [reply src/language-js/parser-babel.js:66]: Flow might recognize those too.

Comment (thorn0) [reply src/language-js/parser-babel.js]: Tested it with Flow. Flow doesn't recognize `wolf@flow`, `+@flow` and even `@flow@` as pragmas. On the other hand, it recognizes `// /@flow` and `// *@flow`. Anyway, let's stay consistent with Babel.

**Review by thorn0 — APPROVED** [review]

**Review by evilebottnawi — APPROVED** [review]

**Comment by fisker** [comment]
Thanks for adding changelog.

**Review by sosukesuzuki — APPROVED** [review]
Looks good!

**Pull request merged**